### ISSUE TITLE
Type deduction fails for `int ** 4`

### ISSUE THREAD

**Issue by ikamensh** [microsoft/pyright#2571]
**Describe the bug**
Pyright fails to deduce type under very specific conditions.

**To Reproduce**
Compare pyright output on the two files below:

Good (no errors): 
```python
from __future__ import annotations

def boo(x: int | None = None) -> int:

    if x is None:
        x = 1
    elif isinstance(x, int):
        x = x ** 2
    else:
        raise RuntimeError
    return x
```

Bad( false error `12 - error: Expression of type "int | None" cannot be assigned to return type "int"`):
```python
from __future__ import annotations

def boo(x: int | None = None) -> int:

    if x is None:
        x = 1
    elif isinstance(x, int):
        x = x ** 4  # changed this
    else:
        raise RuntimeError
    return x
```

The difference is only the power used, 2 vs 4.

**Expected behavior**
In both cases pyright should deduce the type of the expression, as it is well-defined.

**VS Code extension or command-line**
CLI, pyright 1.1.183 and 1.1.187

**Comment by JelleZijlstra** [comment]
This is due to how `__pow__` is defined in typeshed. It returns `Any` because it may return a float if the value is positive. We have a specific overload for `** 2` but not `** 4`. See python/typeshed#6287 for the latest installment in our saga of trying to type `__pow__` better.

**Comment by erictraut** [comment]
As Jelle said, this behavior is due to the way `__pow__` is defined in typeshed. Pyright is applying the type rules correctly.

Here's a (somewhat ugly) workaround:
```python
        x = cast(int, x ** 4)
```
